Let Finders use patient_identifiers

kaapstorm opened the pull request:
This refactors the way we match CommCare cases with OpenMRS patients, so that we can use patient identifiers (like national ID numbers) with PatientFinders the same way we use names, birthdays, etc. 

This allows us to treat a handwritten ID number as "pretty sure it's a match" instead of all-or-nothing. For example, if we have a case in CommCare with the same ID number as a patient in OpenMRS, but they have different names and birthdays, then we don't want to assume that the case is definitely the same person as the patient. If the ID number was an "ID matcher", then CommCare would have matched them, but by using a PatientFinder instead, we have more control of how we match cases with patients.

This PR builds on #18904 -- only the last 4 commits have been added (from "Let Finders use patient_identifiers")

@mkangia cc @dannyroberts buddy @nickpell

kaapstorm committed:
Add examples in comments

kaapstorm committed:
Base class for patient finders

kaapstorm committed:
Define WeightedPropertyPatientFinder

kaapstorm committed:
Define Possible Health patient finder

kaapstorm committed:
Set patient finder in OpenmrsConfig

kaapstorm committed:
Use patient finder if IDs don't match a patient

kaapstorm committed:
Make it obvious that division is py3-compatible

kaapstorm committed:
Save patient ID to case on match

kaapstorm committed:
Logging

kaapstorm committed:
absolute_import

kaapstorm committed:
Support OpenMRS Persion UUID as identifier

kaapstorm committed:
Raise error on bad PatientFinder class name

kaapstorm committed:
Use search_patients()

kaapstorm committed:
Make Finders JSON-configurable

kaapstorm committed:
Protect _property_map so it won't be a JsonDict

kaapstorm committed:
Get patient_value from jsonpath matches

kaapstorm committed:
Let Finders use patient_identifiers

kaapstorm committed:
stickler

kaapstorm committed:
JSONPath expression to compare values

kaapstorm committed:
Better comment

kaapstorm committed:
Tests

kaapstorm committed:
Extend tests

kaapstorm committed:
Use constant, because we have it

kaapstorm committed:
Change "if" to "assert"

kaapstorm committed:
stickler

kaapstorm commented on corehq/motech/openmrs/jsonpath.py:
```diff
@@ -0,0 +1,35 @@
+from jsonpath_rw import JSONPath
+
+
+class WhereCmp(JSONPath):
```
(Ignoring. Implemented `__eq__` to conform to other JSONPath subclasses, but none of them implement `__hash__`.)

mkangia commented on corehq/motech/openmrs/finders.py:
```diff
@@ -0,0 +1,287 @@
+            case_property = case_config_ids[id_type_uuid]['case_property']
+            value = patient['uuid']
+            case_update[case_property] = value
+        for identifier in patient['identifiers']:
```
Are these the identifiers on the patient and now we are saving these on the case?
Would/Should these be blank before this or its fine to override them?

kaapstorm replied:
Yes, they are OpenMRS's identifiers. `case_config_ids` are a subset of them that CommCare has been configured to store in case properties. 

We know that they are blank (or for some reason invalid), because if the case had valid values for at least one of them then the [`get_patient()` function](https://github.com/dimagi/commcare-hq/blob/nh/omrs/identifiers/corehq/motech/openmrs/repeater_helpers.py#L310) would have got the right patient before we had to fall back to searching. 

So this is why we're saving them to the case; so next time we need to match to an OpenMRS patient, we don't have to search; we can match straight away.

mkangia commented on corehq/motech/openmrs/finders.py:
```diff
@@ -0,0 +1,287 @@
+                case.name, case.get_id,
+            )
+            return []
+        patients_scores = sorted(candidates.values(), key=lambda cand: cand.score, reverse=True)
```
We can avoid sorting if `len(patients_scores) == 1`, so can move this after this `if` block?

kaapstorm replied:
lol. Yes, absolutely.

mkangia commented on corehq/motech/openmrs/finders.py:
```diff
@@ -0,0 +1,287 @@
+            'Unable to match case "%s" (%s) to patient candidates: \n%s',
+            case.name, case.get_id, pformat(patients_scores, indent=2),
+        )
+        return [ps.patient for ps in patients_scores]
```
Looks like we are returning `None` if there are more than 1 matches, why are we returning everything then?

kaapstorm replied:
The `find_patient()` function used by the OpenMRS Repeater gives a definitive answer -- either one OpenMRS patient, or `None` if it isn't sure. 

But PatientFinders can be more explicit, and that helps us to know what's going on. So if a PatientFinder doesn't find any patients, it returns an empty list. If it can confidently pick a patient (in the case of this `WeightedPropertyPatientFinder`, because only one passed the threshold, or only one had a score that was much higher than the rest) it must return just that one. But if it found many but can't pick one, we would want to know, and it should return all of them. 

I'm happy to separate the expected behaviour of `PatientFinder.find_patients()` from `repeater_helpers.find_patient()` because a PatientFinder is quite a different idea from the Repeater that uses it.

We might want to write better/different PatientFinders -- other projects might have different requirements. (I'm thinking a [LevenshteinPatientFinder](https://en.wikipedia.org/wiki/Levenshtein_distance) might be useful in some situations.) 

So we want to know if our PatientFinder is consistently letting us down. Finding no patients is fine, because the patient might not exist in OpenMRS, but finding a lot is a bad sign. (We might want to add a soft_assert in `find_patient()` to mail me when we get multiple results.) (... And it would be nice to have a unit test that mocks with copies of real data. Or compare multiple PatientFinder implementations against the same sample set of patients. :thinking: )

mkangia replied:
@kaapstorm That makes sense if this would be used for different patient finders. But if thats not happening soon can consider the overhead of returning all of patients, unless these would always be a small number.

kaapstorm replied:
They really *should* be a small number. Any project that uses PatientFinders will have to watch to see what is being returned, so they can tweak their config. If they are getting a lot of matches, then their weights are wrong.

mkangia reviewed (changes requested):
Some documentation within the code changes would be a bit more helpful.
Left minor possible improvements/suggestions and queries.

kaapstorm committed:
Merge branch 'master' into nh/omrs/identifiers

kaapstorm committed:
Docs

kaapstorm committed:
Raise ValueError on bad ValueSource

kaapstorm committed:
Better comments on JSONPath expressions

kaapstorm committed:
Don't bother sorting list if there is only 1 item

kaapstorm committed:
Simplify jsonpath.WhereCmp to just Cmp

kaapstorm commented:
Hey @mkangia I added some documentation and comments, and took your advice on raising Exceptions, and sorting single-item lists :grin: 

I also simplified my `WhereCmp` jsonpath expression to just `Cmp`.

kaapstorm commented on corehq/motech/openmrs/jsonpath.py:
```diff
@@ -0,0 +1,39 @@
+from __future__ import absolute_import
+
+from operator import gt
```
stickler should read doctests

kaapstorm commented on corehq/motech/openmrs/jsonpath.py:
```diff
@@ -0,0 +1,39 @@
+
+from operator import gt
+
+from jsonpath_rw import JSONPath, Fields
```
Used in doctest

kaapstorm committed:
stickler

mkangia commented:
Thanks @kaapstorm this is easier to understand now ..
